cur-lib/cur/stdlib/pattern-tree.rkt

```diff
@@ -0,0 +1,717 @@
+    (let ([arg (list-ref args idx)])
+      (and (not (list? (syntax->list arg)))
+           (let* ([metadata (get-constructor constructor-id match-var #:env env)]
+                  [arg-binding-types (and metadata (map (compose second syntax->list) (second metadata)))]
+                  [type-parameters (and metadata (third metadata))]
+                  [match-var-type-for-env (and metadata (syntax->list (fourth metadata)))]
+                  [match-var-type-values (and match-var-type-for-env
+                                              (> (length match-var-type-for-env) 2)
+                                              (rest (rest match-var-type-for-env)))]
+                  [match-var-type-bindings (and match-var-type-values (map cons type-parameters match-var-type-values))]
+                  [new-arg-binding-types (and match-var-type-bindings
+                                              (append match-var-type-values
+                                                      (for/list ([ty arg-binding-types])
+                                                        (subst-bindings ty match-var-type-bindings #:equality? (lambda (a b) (equal? (syntax->datum a)
+                                                                                                                                     (syntax->datum b)))))))]
+                  ;; TODO PR103: This is probably wrong now since I changed the
+                  ;; behavior to fail when matching on a type without
+                  ;; constructors.
+                  ;; Doesn't seem to be so that much be some other default case.
+                  ; conveniently, the Type type doesn't have any constructors, so we can use it as the default
+                  [expected-ty (or (and new-arg-binding-types (> (length new-arg-binding-types) idx) (list-ref new-arg-binding-types idx))
+                                   #'Type)]
+                  [dummy-var #'foo])
+             ; case: the argument is a pattern variable if it's in the slot of a generic type
+             ; argument but doesn't correspond to the environment
+             (or (and type-parameters
+                      (< idx (length type-parameters))
+                      (not (equal? (syntax->datum arg)
+                                   (syntax->datum expected-ty))))
+                 ; case: the argument is a pattern variable if it's not a constructor
+                 (not (is-constructor? arg dummy-var #:env (cons (cons dummy-var expected-ty) env))))))))
+
+  ;; Given a pattern of form
+  ;; (C1 (C2 (C3 a1) a2) a3)
+  ;; produces a flat map
+  ;; (#f #t #f)
+  ;; which denotes the position in which new temporaries should be generated
+  (define (generate-tmp-map pattern)
+    (let ([pattern-as-list (syntax->list pattern)])
+      (if (false? pattern-as-list)
+          ;; TODO PR103: Odd that in this case we return a list, which suggest
+          ;; the input was a valid pattern, but it wasn't?
```

Comment:
I think it was a valid pattern, except the form of it isn't list-like. I've used this style a few times and I can see how it's a bit confusing with all the `false?` and `list?` being thrown around.